tests/functional/base.py

```diff
@@ -0,0 +1,662 @@
+import os
+import logging
+import logging.config
+import re
+import time
+import anymarkup
+import datetime
+import unittest
+import subprocess
+from collections import OrderedDict
+import tempfile
+
+from .providers import kubernetes
+from .providers import openshift
+
+LOGGING_CONF = dict(
+    version=1,
+    formatters=dict(
+        bare={
+            "datefmt": "%Y-%m-%d %H:%M:%S",
+            "format": "[%(asctime)s][%(name)10s %(levelname)7s] %(message)s"
+        },
+    ),
+    handlers=dict(
+        console={
+            "class": "logging.StreamHandler",
+            "formatter": "bare",
+            "level": "DEBUG",
+            "stream": "ext://sys.stderr",
+        }
+    ),
+    loggers=dict(
+        test={
+            "level": "DEBUG",
+            "propagate": False,
+            "handlers": ["console"],
+        }
+    )
+)
+
+logging.config.dictConfig(LOGGING_CONF)
+logger = logging.getLogger('test')
+
+
+class BaseProviderTestSuite(unittest.TestCase):
+    """
+    Base test suite for a provider: docker, kubernetes, etc.
+    """
+
+    def setUp(self):
+        self.get_initial_state()
+
+    def tearDown(self):
+        self.restore_initial_state()
+
+    def get_initial_state(self):
+        raise NotImplementedError
+
+    def restore_initial_state(self):
+        raise NotImplementedError
+
+    def deploy(self, app_spec, answers):
+        """
+        Deploy to provider
+        """
+        raise NotImplementedError
+
+    def undeploy(self, app_spec, answers):
+        """
+        Undeploy from provider
+        """
+        raise NotImplementedError
+
+    def get_tmp_answers_file(self, answers):
+        f = tempfile.NamedTemporaryFile(delete=False, suffix='.conf')
+        f.close()
+        anymarkup.serialize_file(answers, f.name, format='ini')
+        return f.name
+
+    @property
+    def nulecule_lib(self):
+        return os.environ.get('NULECULE_LIB') or \
+            os.path.join(os.path.dirname(__file__), '../../../nulecule-library')
```

Comment:
Where do we grab nulecule-library? The original testing script cloned it and built each container.